c10/core/TensorImpl.cpp

```diff
@@ -235,7 +227,6 @@ at::DataPtr PlacementDeleteContext::makeDataPtr(
 AutogradMetaInterface::~AutogradMetaInterface() {}
 
 void TensorImpl::set_requires_grad(bool requires_grad) {
-  TORCH_INTERNAL_ASSERT(type_set_.has(TensorTypeId::VariableTensorId), "set_requires_grad is not implemented for Tensor");
```

Comment:
for these why don't you check that we aren't in the excluded scope?

Reply:
Because it's inconsistent with the old code. The old code only tests `type_set_` and doesn't respect TLS; the new code eliminates the `type_set_` test entirely (because all tensors are variables) and, as before, doesn't respect TLS (which would be the test for excluded scope).

I suppose there is a broader question which is whether or not AutoNonVariableTypeMode should "disable" all autograd operations while it is set. I am not too sure about how to answer this question, but the plan on record is that we make it valid to bypass VariableType entirely for operations which don't require grad. If that becomes the terminal meaning of AutoNonVariableTypeMode (aka `no_grad`), then we absolutely should permit `set_requires_grad` inside a scope like this, in the same way that you can still set requires_grad inside a no_grad block in Python.